fastplotlib/layouts/_engine.py

```diff
@@ -0,0 +1,384 @@
+        # subplot that is in focus, i.e. being hovered by the pointer
+        self._subplot_focus: Subplot | None = None
+
+        for subplot in self._subplots:
+            # highlight plane when pointer enters it
+            subplot.frame.plane.add_event_handler(
+                partial(self._highlight_plane, subplot), "pointer_enter"
+            )
+
+            if resizeable:
+                # highlight/unhighlight resize handler when pointer enters/leaves
+                subplot.frame.resize_handle.add_event_handler(
+                    partial(self._highlight_resize_handler, subplot), "pointer_enter"
+                )
+                subplot.frame.resize_handle.add_event_handler(
+                    partial(self._unhighlight_resize_handler, subplot), "pointer_leave"
+                )
+
+    def _inside_render_rect(self, subplot: Subplot, pos: tuple[int, int]) -> bool:
+        """whether the pos is within the render area, used for filtering out pointer events"""
+        rect = subplot.frame.get_render_rect()
+
+        x0, y0 = rect[:2]
+
+        x1 = x0 + rect[2]
+        y1 = y0 + rect[3]
+
+        if (x0 < pos[0] < x1) and (y0 < pos[1] < y1):
+            return True
+
+        return False
+
+    def canvas_resized(self, canvas_rect: tuple):
+        """
+        called by figure when canvas is resized
+
+        Parameters
+        ----------
+        canvas_rect: (x, y, w, h)
+            the rect that pygfx can render to, excludes any areas used by imgui.
+
+        """
+
+        self._canvas_rect = canvas_rect
+        for subplot in self._subplots:
+            subplot.frame.canvas_resized(canvas_rect)
+
+    def _highlight_resize_handler(self, subplot: Subplot, ev):
+        if self._active_action == "resize":
+            return
+
+        ev.target.material.color = subplot.frame.resize_handle_color.highlight
+
+    def _unhighlight_resize_handler(self, subplot: Subplot, ev):
+        if self._active_action == "resize":
+            return
+
+        ev.target.material.color = subplot.frame.resize_handle_color.idle
+
+    def _highlight_plane(self, subplot: Subplot, ev):
+        if self._active_action is not None:
+            return
+
+        # reset color of previous focus
+        if self._subplot_focus is not None:
+            self._subplot_focus.frame.plane.material.color = (
+                subplot.frame.plane_color.idle
+            )
+
+        self._subplot_focus = subplot
+        ev.target.material.color = subplot.frame.plane_color.highlight
+
+    def __len__(self):
+        return len(self._subplots)
+
+
+class FlexLayout(BaseLayout):
+    def __init__(
+        self,
+        renderer,
+        subplots: np.ndarray[Subplot],
+        canvas_rect: tuple,
+        moveable=True,
+        resizeable=True,
+    ):
+        """
+        Flexible layout engine that allows freely moving and resizing subplots.
+        Subplots are not allowed to overlap.
```

Comment:
The name `FlexLayout` makes me think immediately about CSS flexbox, which is a common layout method. But this seems to do something else (allowing users to freely move the contents). You could consider a different name to avoid confusion.

Reply:
changed to `WindowLayout`